File: services/report-service/src/main/java/com/otterworks/report/util/ReportDateUtils.java
package com.otterworks.report.util;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.time.DateFormatUtils;
import org.apache.commons.lang3.time.DateUtils;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

/**
 * Date utility class for report formatting and parsing.
 */
public final class ReportDateUtils {

    private static final SimpleDateFormat ISO_FORMAT = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
    private static final SimpleDateFormat DISPLAY_FORMAT = new SimpleDateFormat("MMM dd, yyyy HH:mm");
    private static final SimpleDateFormat FILE_NAME_FORMAT = new SimpleDateFormat("yyyyMMdd_HHmmss");

    static {
        ISO_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
        DISPLAY_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
        FILE_NAME_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
    }

    private ReportDateUtils() {
        // Utility class
    }

    /**
     * Format a Date to ISO 8601 string.
     */
    public static String toIsoString(Date date) {
        if (date == null) {
            return null;
        }
        return DateFormatUtils.formatUTC(date, "yyyy-MM-dd'T'HH:mm:ss'Z'");
    }

    /**
     * Format a Date for display purposes.
     */
    public static String toDisplayString(Date date) {
        if (date == null) {
            return "N/A";
        }
        synchronized (DISPLAY_FORMAT) {
            return DISPLAY_FORMAT.format(date);
        }
    }

    /**
     * Format a Date for use in file names.
     */
    public static String toFileNameString(Date date) {
        if (date == null) {
            date = new Date();
        }
        synchronized (FILE_NAME_FORMAT) {
            return FILE_NAME_FORMAT.format(date);
        }
    }

    /**
     * Parse an ISO 8601 date string.
     */
    public static Date parseIsoDate(String dateString) {
        if (StringUtils.isBlank(dateString)) {
            return null;
        }
        try {
            return DateUtils.parseDate(dateString, new String[]{
                    "yyyy-MM-dd'T'HH:mm:ss'Z'",
                    "yyyy-MM-dd'T'HH:mm:ssZ",
                    "yyyy-MM-dd HH:mm:ss",
                    "yyyy-MM-dd"
            });
        } catch (ParseException e) {
            throw new IllegalArgumentException("Cannot parse date: " + dateString, e);
        }
    }

    /**
     * Get the start of today (midnight UTC).
     */
    public static Date startOfToday() {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        return cal.getTime();
    }

    /**
     * Get the start of this month.
     */
    public static Date startOfMonth() {
        Calendar cal = Calendar.getInstance(TimeZone.getTimeZone("UTC"));
        cal.set(Calendar.DAY_OF_MONTH, 1);
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        return cal.getTime();
    }

    /**
     * Subtract days from a date.
     */
    public static Date daysAgo(int days) {
        return DateUtils.addDays(new Date(), -days);
    }

    /**
     * Check if a date falls within a range (inclusive).
     */
    public static boolean isWithinRange(Date date, Date start, Date end) {
        if (date == null || start == null || end == null) {
            return false;
        }
        return !date.before(start) && !date.after(end);
    }

    /**
     * Calculate duration between two dates in a human-readable format.
     */
    public static String humanReadableDuration(Date start, Date end) {
        if (start == null || end == null) {
            return "unknown";
        }
        long diffMs = end.getTime() - start.getTime();
        long seconds = diffMs / 1000;
        long minutes = seconds / 60;
        long hours = minutes / 60;

        if (hours > 0) {
            return hours + "h " + (minutes % 60) + "m";
        } else if (minutes > 0) {
            return minutes + "m " + (seconds % 60) + "s";
        } else {
            return seconds + "s";
        }
    }
}

File: services/report-service/Utilities/ReportDateUtils.cs
using System.Globalization;

namespace OtterWorks.ReportService.Utilities;

public static class ReportDateUtils
{
    private static readonly string[] IsoFormats =
    [
        "yyyy-MM-dd'T'HH:mm:ss'Z'",
        "yyyy-MM-ddTHH:mm:ssZ",
        "yyyy-MM-ddTHH:mm:sszzz",
        "yyyy-MM-dd HH:mm:ss",
        "yyyy-MM-dd",
    ];

    public static string? ToIsoString(DateTime? date)
    {
        if (date == null)
        {
            return null;
        }

        return date.Value.ToUniversalTime().ToString("yyyy-MM-dd'T'HH:mm:ss'Z'", CultureInfo.InvariantCulture);
    }

    public static string ToDisplayString(DateTime? date)
    {
        if (date == null)
        {
            return "N/A";
        }

        return date.Value.ToUniversalTime().ToString("MMM dd, yyyy HH:mm", CultureInfo.InvariantCulture);
    }

    public static string ToFileNameString(DateTime? date)
    {
        var d = date ?? DateTime.UtcNow;
        return d.ToUniversalTime().ToString("yyyyMMdd_HHmmss", CultureInfo.InvariantCulture);
    }

    public static DateTime ParseIsoDate(string dateString)
    {
        if (string.IsNullOrWhiteSpace(dateString))
        {
            throw new ArgumentException("Date string cannot be null or empty", nameof(dateString));
        }

        if (DateTime.TryParseExact(
            dateString,
            IsoFormats,
            CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
            out var result))
        {
            return result;
        }

        throw new ArgumentException($"Cannot parse date: {dateString}", nameof(dateString));
    }

    public static DateTime StartOfToday()
    {
        return DateTime.UtcNow.Date;
    }

    public static DateTime StartOfMonth()
    {
        var now = DateTime.UtcNow;
        return new DateTime(now.Year, now.Month, 1, 0, 0, 0, DateTimeKind.Utc);
    }

    public static DateTime DaysAgo(int days)
    {
        return DateTime.UtcNow.AddDays(-days);
    }

    public static bool IsWithinRange(DateTime date, DateTime start, DateTime end)
    {
        return date >= start && date <= end;
    }

    public static string HumanReadableDuration(DateTime start, DateTime end)
    {
        var diff = end - start;
        long totalSeconds = (long)diff.TotalSeconds;
        long minutes = totalSeconds / 60;
        long hours = minutes / 60;

        if (hours > 0)
        {
            return $"{hours}h {minutes % 60}m";
        }

        if (minutes > 0)
        {
            return $"{minutes}m {totalSeconds % 60}s";
        }

        return $"{totalSeconds}s";
    }
}
